orc8r/gateway/python/magma/ctraced/trace_manager.py

```diff
@@ -0,0 +1,117 @@
+limitations under the License.
+"""
+
+import logging
+import errno
+import os
+import pathlib
+import subprocess
+from subprocess import SubprocessError
+from .command_builder import get_trace_builder
+
+_TRACE_FILE_NAME = "call_trace.pcap"
+_MAX_FILESIZE = 4000  # ~ 4 MiB for a trace
+
+
+class TraceManager:
+    """
+    TraceManager is a wrapper for tshark/tcpdump specifically for starting and
+    stopping call/interface/subscriber traces.
+
+    Only a single trace can be captured at a time.
+    """
+    def __init__(self, config):
+        self._is_active = False # is call trace being captured
+        self._proc = None
+        self._trace_directory = config.get("trace_directory",
+                                           "/var/opt/magma/trace")  # type: str
+        # Specify southbound interfaces
+        self._trace_interfaces = config.get("trace_interfaces",
+                                           ["eth0"])  # type: List[str]
+        self._trace_filename = self._trace_directory + "/" + _TRACE_FILE_NAME
+
+        tool_name = config.get("trace_tool", "tshark")  # type: str
+        self._trace_builder = get_trace_builder(tool_name)
+
+    def start_trace(self) -> bool:
+        """Start a call trace.
+
+        Captures all packets across the eth0 interface.
+
+        Returns:
+            True if successfully started call trace
+        """
+        if self._is_active:
+            logging.error("Failed to start trace: Trace already active")
+            return False
+
+        command = self._trace_builder.build_trace_command(
+            self._trace_interfaces,
+            _MAX_FILESIZE,
+            self._trace_filename)
+
+        logging.info("Starting trace with tshark, command: [%s]",
+                     ' '.join(command))
+
+        self._ensure_trace_directory_exists()
+
+        # TODO(andreilee): Handle edge case where only one instance of the
+        #                  process can be running, and may have been started
+        #                  by something external as well.
+        try:
+            self._proc = subprocess.Popen(command)
+        except SubprocessError as e:
+            logging.error("Failed to start trace: %s", str(e))
+            return False
+
+        self._is_active = True
+        logging.info("Successfully started trace with tshark")
+        return True
+
+    def end_trace(self) -> bytes:
+        """Ends call trace, if currently active.
+
+        Returns:
+            Call trace file in bytes
+        """
+        # If trace is active, then stop it
+        if self._is_active:
+            # If the process has ended, then _proc isn't None
+            self._proc.poll()
+            if self._proc.returncode is None:
+                self._proc.terminate()
+
+        # Read trace data into bytes
+        with open(self._trace_filename, "rb") as trace_file:
+            data = trace_file.read()  # type: bytes
+
+        self._ensure_tmp_file_deleted()
```

Comment:
Might be better to return the call tracing data and then throw a logging error when a temp file cannot be deleted. Call tracing service failing due to this error in middle of user debugging session might not be desirable.

Reply:
Sure